### ISSUE TITLE
pulling an empty repository

### ISSUE THREAD

**Issue by hannesm** [mirage/irmin#671]
My code is a MirageOS unikernel that pulls a remote and prints out the top-level bindings found (waits for 5 seconds and pulls again) -- code at https://gist.github.com/hannesm/295c0d65dae3f7a817d3378ccaf623f7 -- not sure where the source of this issue is (either in git or in irmin).

Now, if I use as a remote an empty repository (such as https://github.com/hannesm/empty.git) either via the git or https protocol (both show the same behaviour), I get a fetch error (I expected an Ok from pull) -- this uses ocaml-git from master, where https://github.com/mirage/ocaml-git/pull/351 is merged, and irmin including #539
```
2019-04-10 14:54:56 -00:00: INF [application] i have now master!
2019-04-10 14:54:56 -00:00: INF [application] pulling from remote!
2019-04-10 14:54:56 -00:00: DBG [irmin.sync] Fetching branch master
2019-04-10 14:54:56 -00:00: DBG [irmin.git] fetch https://github.com/hannesm/empty.git
2019-04-10 14:54:56 -00:00: DBG [git.mem] Ref.list
2019-04-10 14:54:56 -00:00: DBG [negociator] Local references: .
2019-04-10 14:54:56 -00:00: DBG [git.sync.http] Launch the GET request to https://github.com/hannesm/empty.git/info/refs?service=git-upload-pack.
2019-04-10 14:54:56 -00:00: DBG [git.sync.http] Send the GET (reference discovery) request.
2019-04-10 14:54:56 -00:00: DBG [cohttp] Send a request to https://github.com/hannesm/empty.git/info/refs?service=git-upload-pack.
2019-04-10 14:54:56 -00:00: DBG [irmin.git] find master
2019-04-10 14:54:56 -00:00: DBG [git.mem] Ref.read refs/heads/master
2019-04-10 14:54:56 -00:00: DBG [irmin] Head.find -> 
2019-04-10 14:54:56 -00:00: WRN [application] fetch error empty head!
```

**Comment by zshipko** [comment]
This would be an easy fix (I have it here: https://github.com/zshipko/irmin/tree/fix-pull-empty), but after thinking about it some more, maybe the `No_head` error is actually information that should be preserved. In your example you could easily check for it in the `pull_store` function - I'm on the fence about this!

**Comment by hannesm** [comment]
@zshipko oh, indeed there's a nice tag for it. I'm a bit on the fence about this `pull` function: it returns `Ok ()` on success, then ``Error `No_head`` if you cloned an empty repository, and ``Error `Not_available`` on a real failure (not found? what about remote endpoint not answering?). also: ``Error `Conflict of string`` is possible for merge conflicts, and ``Error `Msg of string`` which is not very clear to me when this is signalled. Maybe instead of modifying the code, the docstring should be improved (and if there are so many (nice!) error tags, why/when is `` `Msg `` used? can that be avoided?).

it reminds me of ``type 'a or_eof = [`Data of 'a | `Eof ]`` (in [mirage-flow](https://github.com/mirage/mirage-flow/blob/v1.5.0/src/mirage_flow.mli#L31)) used as left/Ok return type of `read`, which I don't like too much -- the `` `No_head `` / `` `Eof`` is for most applications not an error case, but making the left/Ok result more complex for this sake feels wrong... I guess from a user point of view, I'd like to have a tri-state result:
- ok, I pulled the remote you mentioned
- I pulled, but its empty, no point in doing anything with your data now
- pull returned an error (which you may solve (i.e. merge conflict, discard my changes) or not (i.e. protocol failure, network connection problem, authentication problem))

**Comment by samoht** [comment]
It makes sense to have a three-state result for pull.

**Comment by hannesm** [comment]
@samoht I'm not entirely sure anymore. the current API is fine as well, I guess what is missing is some documentation about the errors (and a pretty-printer, as remarked somewhere else on github)... whether or not to treat `` `Empty_head `` as ok or error does not matter too much imho

**Comment by zshipko** [comment]
I will definitely add the pretty printers and some documentation about the errors today.

**Comment by hannesm** [comment]
this seems to be solved by #684, thanks
